### PR TITLE
[Material] Adaptive Slider constructor

### LINKED ISSUE #30571: Material Slider has no adaptive constructor
The switch has an adaptive constructor. The slider should also have one.

### PR DESCRIPTION
**Pull request by clocksmith**
## Description

Adds an adaptive constructor for the Material Slider. An adaptive widget is one that renders itself as Material on Android, and Cupertino on iOS. This work is based off of a similar feature on Switches: https://github.com/flutter/flutter/commit/bbb080b3a3d4e2f3b61b2f7b029cafdb9b9cab99#diff-fe2bb980c6207699cbf45538fe927afa.

The motivation for this change is that we should provide adaptive constructors for as many widgets as necessary in the Material library. In Material, it is suggested that the slider is an iOS-style slider.

## Related Issues

closes #30571 

## Tests

I added the following tests:

Slider.adaptive, which builds an adaptive Slider, and changes the platform checking for the expected resulting Widget.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/slider.dart [hunk 1]
```diff
@@ -121,7 +124,37 @@ class Slider extends StatefulWidget {
...
+       assert(value != null),
+       assert(min != null),
+       assert(max != null),
+       assert(min <= max),
+       assert(value >= min && value <= max),
+       assert(divisions == null || divisions > 0),
+       super(key: key);
+
+  /// Creates a [CupertinoSlider] if the target platform is iOS, creates a
+  /// material design slider otherwise.
```
Comment (willlarche) [inline packages/flutter/lib/src/material/slider.dart]: `Material Design`

File: packages/flutter/lib/src/material/slider.dart [hunk 2]
```diff
@@ -433,6 +474,27 @@ class _SliderState extends State<Slider> with TickerProviderStateMixin {
     assert(debugCheckHasMaterial(context));
     assert(debugCheckHasMediaQuery(context));
 
+    switch (widget._sliderType) {
+      case _SliderType.material:
+        return _buildMaterialSlider(context);
+
+      case _SliderType.adaptive: {
+        final ThemeData theme = Theme.of(context);
```
Comment (willlarche) [inline packages/flutter/lib/src/material/slider.dart]: Are we guaranteed a Theme if it's a CupertinoApp?

File: packages/flutter/lib/src/material/slider.dart [hunk 3]
```diff
@@ -463,6 +525,25 @@ class _SliderState extends State<Slider> with TickerProviderStateMixin {
       semanticFormatterCallback: widget.semanticFormatterCallback,
     );
   }
+
+  Widget _buildCupertinoSlider(BuildContext context) {
+    // The render box of a slider has a fixed height by takes up the available
```
Comment (willlarche) [inline packages/flutter/lib/src/material/slider.dart]: Check grammar.

File: packages/flutter/lib/src/material/slider.dart [hunk 4]
```diff
@@ -463,6 +525,25 @@ class _SliderState extends State<Slider> with TickerProviderStateMixin {
...
+    // width. Wrapping the cupertino slider in this manner will help ensure
```
Comment (willlarche) [inline packages/flutter/lib/src/material/slider.dart]: [CupertinoSlider]. Might be others in this comment.

**Review by johnsonmh — APPROVED** [review]
LGTM

File: packages/flutter/test/material/slider_test.dart [hunk 5]
```diff
@@ -1422,4 +1423,50 @@ void main() {
...
+    expect(value, 0.5);
+    TestGesture gesture = await tester.startGesture(tester.getCenter(find.byType(CupertinoSlider)));
+    // Drag to the right end of the track.
+    await gesture.moveBy(const Offset(600.0, 0.0));
+    expect(value, 1.0);
+
+    value = 0.5;
+    await tester.pumpWidget(buildFrame(TargetPlatform.android));
+    await tester.pumpAndSettle(); // Finish the theme change animation.
+    expect(find.byType(CupertinoSlider), findsNothing);
```
Comment (willlarche) [inline packages/flutter/test/material/slider_test.dart]: Maybe have a does find Slider.

**Comment by willlarche** [comment]
LGTM

File: packages/flutter/lib/src/material/slider.dart (hunk 2 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/slider.dart]: Yes Theme.of always returns a fallback theme at the least

File: packages/flutter/test/material/slider_test.dart (hunk 5 above)
Comment (clocksmith) [reply packages/flutter/test/material/slider_test.dart]: Done

File: packages/flutter/lib/src/material/slider.dart (hunk 4 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/slider.dart]: Done

File: packages/flutter/lib/src/material/slider.dart (hunk 1 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/slider.dart]: Done and elsewhere in this file

File: packages/flutter/lib/src/material/slider.dart [hunk 6]
```diff
@@ -121,7 +124,37 @@ class Slider extends StatefulWidget {
     this.activeColor,
     this.inactiveColor,
     this.semanticFormatterCallback,
-  }) : assert(value != null),
+  }) : _sliderType = _SliderType.material,
+       assert(value != null),
```
Comment (rami-a) [inline packages/flutter/lib/src/material/slider.dart]: We may want to document these assertions in the constructor documentation.

File: packages/flutter/lib/src/material/slider.dart [hunk 7]
```diff
@@ -121,7 +124,37 @@ class Slider extends StatefulWidget {
...
+    this.onChangeEnd,
+    this.min = 0.0,
+    this.max = 1.0,
+    this.divisions,
+    this.label,
+    this.activeColor,
+    this.inactiveColor,
+    this.semanticFormatterCallback,
+  }) : _sliderType = _SliderType.adaptive,
+       assert(value != null),
```
Comment (rami-a) [inline packages/flutter/lib/src/material/slider.dart]: We may want to document these assertions in the constructor documentation.

**Review by rami-a — APPROVED** [review]
Looks good, just a small comment

File: packages/flutter/lib/src/material/slider.dart (hunk 3 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/slider.dart]: Done

File: packages/flutter/lib/src/material/slider.dart (hunk 6 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/slider.dart]: These changes are based off switch.dart, which does not document the assertions in the ctor docs. If they are supposed to be there, I can make a bug to add docs to both classes.

File: packages/flutter/lib/src/material/slider.dart (hunk 7 above)
Comment (clocksmith) [reply packages/flutter/lib/src/material/slider.dart]: see comment above

**Review by willlarche — APPROVED** [review]

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**